Use Vec directly in generic_bytes_builder

liamzwbao opened the pull request:
# Which issue does this PR close?

We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax.

- Closes #NNN.

# Rationale for this change

Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.

# What changes are included in this PR?

There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.

# Are these changes tested?

We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?

# Are there any user-facing changes?

If there are user-facing changes then we may require documentation to be updated before approving the PR.

If there are any breaking changes to public APIs, please call them out.

liamzwbao commented:
Hi @alamb and @Dandandan, this PR is ready for review. I have changed the implementation to use `Vec` in a few builders. Not sure if some of them are not appropriate to migrate to this new implementation.

Have tested the performance for arrow-array locally, half improved half regressed.

liamzwbao commented on arrow-array/src/builder/primitive_builder.rs:
```diff
@@ -296,7 +297,7 @@ impl<T: ArrowPrimitiveType> PrimitiveBuilder<T> {
 
         self.null_buffer_builder.append_n_non_nulls(len);
-        self.values_builder.append_trusted_len_iter(iter);
+        self.values_builder.extend(iter);
```
There is no 100% matched function for `Vec` that will check the upper bound of the iter. But this function already check the upper bound of size_hint on line 294, I think it's okay to replace it with `extend` directly

jhorstmann replied:
If the `iter` actually implements the stdlib internal [`TrustedLen`](https://doc.rust-lang.org/std/iter/trait.TrustedLen.html) trait then this should get nicely optimized. The only caller inside the codebase seems to be in `arrow-select` with a Range iterator, which implements that trait.

Dandandan commented on arrow-array/src/builder/generic_bytes_builder.rs:
```diff
@@ -67,8 +67,8 @@ impl<T: ByteArrayType> GenericByteBuilder<T> {
-        let offsets_builder = BufferBuilder::<T::Offset>::new_from_buffer(offsets_buffer);
-        let value_builder = BufferBuilder::<u8>::new_from_buffer(value_buffer);
+        let offsets_builder = Vec::from(offsets_buffer.typed_data::<T::Offset>());
+        let value_builder = Vec::from(value_buffer.as_slice());
```
This seems it does a copy?

liamzwbao replied:
Replace it with ScalarBuffer instead

Dandandan commented on arrow-array/src/builder/generic_bytes_builder.rs:
```diff
@@ -126,15 +127,16 @@ impl<T: ByteArrayType> GenericByteBuilder<T> {
         let next_offset = self.next_offset();
-        self.offsets_builder.append_n(n, next_offset);
+        self.offsets_builder
+            .extend(std::iter::repeat(next_offset).take(n));
```
`repeat_n`?

Dandandan commented on arrow-array/src/builder/generic_bytes_builder.rs:
```diff
@@ -184,11 +186,11 @@ impl<T: ByteArrayType> GenericByteBuilder<T> {
             .len(self.len())
-            .add_buffer(self.offsets_builder.finish())
-            .add_buffer(self.value_builder.finish())
+            .add_buffer(Buffer::from_vec(std::mem::take(&mut self.offsets_builder)))
```
Should be able to use `into()`

Dandandan commented on arrow-array/src/builder/primitive_builder.rs:
```diff
@@ -162,7 +162,7 @@ impl<T: ArrowPrimitiveType> PrimitiveBuilder<T> {
         null_buffer: Option<MutableBuffer>,
     ) -> Self {
-        let values_builder = BufferBuilder::<T::Native>::new_from_buffer(values_buffer);
+        let values_builder = Vec::from(values_buffer.typed_data::<T::Native>());
```
Does a copy?

liamzwbao commented:
Hi @jhorstmann @Dandandan, thank you very much for the review! I have addressed the comments.

However, I noticed that this implementation will fail some tests because the capacity of `Vec` does not round up to multiple of 64 like [MutableBuffer](https://github.com/apache/arrow-rs/blob/16794ab14fa62ecf67de0da9460cc5752a9358f4/arrow-buffer/src/buffer/mutable.rs#L88) does. I have updated the tests in this PR for reference. Would appreciate suggestions to address this issue.

alamb commented:
Are there any benchmarks that would be good for me to run for this PR?

liamzwbao commented:
> Are there any benchmarks that would be good for me to run for this PR?

I think arrow-array is a good candidate

alamb commented on arrow-select/src/concat.rs:
```diff
@@ -1328,12 +1328,12 @@ mod tests {
         let data = a.to_data();
         assert_eq!(data.buffers()[0].len(), 440);
-        assert_eq!(data.buffers()[0].capacity(), 448); // Nearest multiple of 64
+        assert_eq!(data.buffers()[0].capacity(), 440); // Nearest multiple of 64
```
these comments no longer seem correct, but the tests seem to be an improvement

alamb reviewed (approved):
THank you @liamzwbao @Dandandan and @jhorstmann -- this PR looks good to me

I queued up some benchmarks to see if we can see any improvements

alamb commented:
🤖: Benchmark completed

<details><summary>Details</summary>
<p>


```
group                              issue-7383                             main
-----                              ----------                             ----
array_from_vec 128                 1.00    152.0±0.34ns        ? ?/sec    1.01    153.8±0.30ns        ? ?/sec
array_from_vec 256                 1.00    161.0±1.43ns        ? ?/sec    1.01    162.6±0.26ns        ? ?/sec
array_from_vec 512                 1.00    212.4±0.41ns        ? ?/sec    1.00    212.6±0.23ns        ? ?/sec
array_string_from_vec 128          1.00   1115.8±1.52ns        ? ?/sec    1.09   1217.6±2.62ns        ? ?/sec
array_string_from_vec 256          1.00   1939.2±2.78ns        ? ?/sec    1.07      2.1±0.00µs        ? ?/sec
array_string_from_vec 512          1.00      3.3±0.00µs        ? ?/sec    1.02      3.4±0.01µs        ? ?/sec
decimal128_array_from_vec 32768    1.00     99.2±0.50µs        ? ?/sec    1.00     99.1±0.22µs        ? ?/sec
decimal256_array_from_vec 32768    1.00      3.8±0.02µs        ? ?/sec    1.00      3.9±0.02µs        ? ?/sec
struct_array_from_vec 1024         1.01      8.8±0.02µs        ? ?/sec    1.00      8.7±0.02µs        ? ?/sec
struct_array_from_vec 128          1.02   1941.8±5.96ns        ? ?/sec    1.00   1894.6±3.08ns        ? ?/sec
struct_array_from_vec 256          1.00      2.9±0.01µs        ? ?/sec    1.03      2.9±0.01µs        ? ?/sec
struct_array_from_vec 512          1.00      4.8±0.01µs        ? ?/sec    1.00      4.8±0.01µs        ? ?/sec
```


</p>
</details>

alamb commented:
Thanks again @liamzwbao and @Dandandan and @jhorstmann
